### DISCUSSION TITLE
How Can I Install ARM versions of Python?

### DISCUSSION THREAD

**Discussion by russ-mcbride** [pyenv/pyenv#2880] [Q&A] [answered]
Related to #2878--

I am able to do python installs now with a call to, e.g., `arch -arm64 pyenv install 3.11.7` but when I check the python kernel using `import platform; platform.machine()` I get 'x86_64'. 

Why am I getting x86 python versions when brew, openssl, and readline are all arm-native and I'm calling pyenv with an 'arch -arm64 prefix?

Note that the only way I could get python builds to work was by removing readline, and openssl from Homebrew, thereby forcing the download of them during the install. openssl 3 and readline 8.2 are both arm-native.

**Comment by native-api** [comment]
Need diagnostic information as per the issue template to say anything specific

**Comment by russ-mcbride** [comment]
pyenv 2.3.35
Homebrew 4.2.2
OS Sonoma 14.2.1

The new installs are quite terse:
`pyenv install 3.12.1
Downloading openssl-3.1.2.tar.gz...
-> https://www.openssl.org/source/openssl-3.1.2.tar.gz
Installing openssl-3.1.2...
Installed openssl-3.1.2 to /Users/russ/.pyenv/versions/3.12.1
Downloading readline-8.2.tar.gz...
-> https://ftpmirror.gnu.org/readline/readline-8.2.tar.gz
Installing readline-8.2...
Installed readline-8.2 to /Users/russ/.pyenv/versions/3.12.1
Downloading Python-3.12.1.tar.xz...
-> https://www.python.org/ftp/python/3.12.1/Python-3.12.1.tar.xz
Installing Python-3.12.1...
python-build: use tcl-tk from homebrew
python-build: use zlib from xcode sdk
Installed Python-3.12.1 to /Users/russ/.pyenv/versions/3.12.1`

I'm happy to provide more informative diagnostics if you can tell me what to grab. :-)


P.S. I just installed pyenv-virtualenv which seems to have triggered the install of openssl and other libraries, and of course I'm getting the 'no module named _ssl' error again.

**Reply by native-api** [reply]
check the issue template

**Reply by russ-mcbride** [reply]
Forgive my ignorance, where is it located?

**Reply by native-api** [reply]
when you're creating a new issue

**Comment by russ-mcbride** [comment]
### Description
Mac OS Sonoma 14.2.1
M1 (Apple Silicon) Pro Max
Pyenv 2.3.35
Python Install 3.9.18
Apple Clang version 15.0.0 (clang-1500.1.0.2.5)
[trace.log](https://github.com/pyenv/pyenv/files/13834669/trace.log)
[config.log](https://github.com/pyenv/pyenv/files/13834667/config.log)

**Reply by native-api** [reply]
Your active architecture is x64:

```
uname -m = x86_64
<...>
configure:3910: clang --version >&5
Apple clang version 15.0.0 (clang-1500.1.0.2.5)
Target: x86_64-apple-darwin23.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

You have probably run the current shell with `arch -x86_64`.

**Reply by russ-mcbride** [reply]
Thanks.

But when I run "arch" in my zsh shell I get: arm64. 
When I run uname -m in my shell I get: arm64
And my 'pyenv' is aliased to 'arch -arm64 pyenv.

I'm confused. Am I missing something obvious?

**Reply by native-api** [reply]
Looks like that arch is getting forced somehow:

> clang  -I. -Iinclude -fPIC **-arch x86_64** -O3 -Wall -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DAESNI_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/Users/russ/.pyenv/versions/3.9.18/openssl/ssl\"" -DENGINESDIR="\"/Users/russ/.pyenv/versions/3.9.18/openssl/lib/engines-1.1\"" -D_REENTRANT -DZLIB -DZLIB_SHARED -DNDEBUG -I/Users/russ/.pyenv/versions/3.9.18/include -MMD -MF apps/app_rand.d.tmp -MT apps/app_rand.o -c -o apps/app_rand.o apps/app_rand.c

Check your environment and shell configuration files for where that `-arch x86_64` comes from.

**Comment by russ-mcbride** [comment]
This is my entire .zshrc:

```
setopt no_nomatch
echo "Now reading .zshrc..."

alias arm="env /usr/bin/arch -arm64 /bin/zsh --login"
alias intel="env /usr/bin/arch -x86_64 /bin/zsh --login"

alias brew86="arch -x86_64 /usr/local/bin/brew"
alias brew="arch -arm64 /opt/homebrew/bin/brew"

alias pyenv86="arch -x86_64 pyenv"
alias pyenv="arch -arm64 pyenv"


#For pyenv:
export PYENV_ROOT="$HOME/.pyenv"
export PATH="$PYENV_ROOT/bin:$PATH"
eval "$(pyenv init --path)"
eval "$(pyenv init -)"
```

I don't see how either of the x86 aliases would cause the problem is they're not being called. Unless something is getting added somewhere as a result of the upgrade to Sonoma...  ??

**Reply by native-api** [reply]
Look at your envvars and elsewhere on the disk. It must be coming from somewhere. E.g. previously, we had a case of it coming from the ARCHFLAGS envvar which was set for some other software.

**Comment by russ-mcbride** [comment, marked as answer]
SOLVED! I lost 8 days of my life but learned a few things. Somehow, the x86 version of brew was getting engaged when I'd call the arm version with:  arch -arm64 /opt/homebrew/bin/brew. I deleted the x86 Homebrew installation and now the arm python install works flawlessly. I reinstalled x86 Homebrew and and now I can install either arm versions of python or x86 versions of python as I want. It appears that the first secret is to install the arm version of Homebrew first (at least that's what worked for me). The second secret is force delete the openssl and readline formulae from x86 brew library otherwise I can't install anything on the x86 side.

I'm happy to offer further details on my setup if anyone wants it....

**Reply by varenc** [reply]
Thanks for this tip! I just ran into the same issue.  My old x86 homebrew was in `/usr/local/Cellar/`.  I just moved it to `/usr/local/Cellar_DISABLED` and then the `pyenv install` worked and built an arm64 version.  Moving Cellar back to its original location after let me restore my x86 homebrew things without uninstalling it all completely.  (not exactly good practice though)

**Reply by gkomail-afk** [reply]
Thank you so much brother, it solved the issue having from months. Thanks again.
